x/programs/rust/examples/cpmm/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+use wasmlanche_sdk::{public, state_keys, Context, Program};
+
+#[state_keys]
+pub enum StateKeys {
+    // Internal accounting
+    ReserveX,
+    ReserveY,
+
+    // Liquidity token
+    TotalySupply,
+}
+
+#[public]
+pub fn add_liquidity(context: Context<StateKeys>, amount_x: u64, amount_y: u64) -> u64 {
+    let program = context.program();
+    let total_supply = total_supply(program);
+    // tokens    | shares
+    // -------------------------
+    // amount_x  | minted
+    // reserve_x | total_supply
+    let (reserve_x, reserve_y) = reserves(program);
+    let minted = if total_supply == 0 {
+        let minted = amount_x;
+        assert_eq!(minted, amount_y);
+        minted
+    } else {
+        let minted = amount_x * total_supply / reserve_x;
+        assert_eq!(minted, amount_y * total_supply / reserve_y); // make sure that the ratio is good
+        minted
+    };
+    program
+        .state()
+        .store(StateKeys::ReserveX, &(reserve_x + amount_x))
+        .unwrap();
+    program
+        .state()
+        .store(StateKeys::ReserveY, &(reserve_y + amount_y))
+        .unwrap();
+    program
+        .state()
+        .store(StateKeys::TotalySupply, &(total_supply + minted))
+        .unwrap();
+    minted
+}
+
+#[public]
+pub fn remove_liquidity(context: Context<StateKeys>, shares: u64) -> (u64, u64) {
+    let program = context.program();
+    let total_supply = total_supply(program);
+    let (reserve_x, reserve_y) = reserves(program);
+    let (amount_x, amount_y) = (
+        shares * reserve_x / total_supply,
+        shares * reserve_y / total_supply,
+    );
+    program
+        .state()
+        .store(StateKeys::ReserveX, &(reserve_x - amount_x))
+        .unwrap();
+    program
+        .state()
+        .store(StateKeys::ReserveY, &(reserve_y - amount_y))
+        .unwrap();
+    program
+        .state()
+        .store(StateKeys::TotalySupply, &(total_supply - shares))
+        .unwrap();
+    (amount_x, amount_y)
+}
+
+#[public]
+pub fn swap(context: Context<StateKeys>, amount_in: u64, x_to_y: bool) -> u64 {
```

Comment:
You didn't want to just do this where you have `amount_x` and `amount_y` in? Another option is to use an enum rather than a bool. This is Rust after all...

Reply:
For `add_liquidity` ? We probably don't want that because the liquidity has to be added on both side and with a similar ratio for each token.

Reply:
Can use an enum though for `swap`, but we currently don't support these in the simulator

Reply:
I meant for swap. But I see that the simulator doesn't support it, that's too bad. The simulator should be able to support anything that's serialized shouldn't it? You want to fix that?

Reply:
Let's fix that !